### ISSUE TITLE
Non-HTML Templates

### ISSUE THREAD

**Issue by laurence-myers** [haxetink/tink_template#3]
Hi Juraj, I love your Haxe libraries, they're all really fantastic, and show how much can be achieved at compile-time! I'm currently using your tink_template library for a fairly straightforward code generator tool.

How hard would it be to enable non-HTML templates? I'm happy to make the change myself. I wanted to know how you would like to expose this option in the API. Is it possible to declare the return type of the template functions, and inspect that return type in the macro to work out what escaping to use?

```
@:template static public function renderTownHtml(t:Town) : Html;
@:template static public function renderTownText(t:Town) : String;
```

This could possibly allow users to specify their own return types (and hence escaping rules). I'm not sure if the macro would need to know of each return type, or if each type would need to be an abstract over String, or if the macro can do some magic with creating an instance of the declared return type.

**Comment by back2dos** [comment]
Hey Laurence, thanks for your kind words!

I indeed [have plans for this](https://github.com/haxetink/tink_template#other-formats) but never really got around to it.

There's many ways to go about this. One would be to use the return type, another to use the file extension. This week I'm rather busy preparing for the WWX, so my mind is too preoccupied to form a proper opinion right now, but any suggestions are highly welcome ;)

**Comment by laurence-myers** [comment]
No problem, good luck with WWX! I look forward to watching the videos online. :)

Regarding using file extensions, I see they are currently used to distinguish between 'mtt' and 'tt' syntaxes and features (like 'foreach'). If we want to preserve this functionality, the template might need dual extensions, like 'template.html.tt'. There'd also need to be a way to register extensions with the class/function/abstract that provides escaping, and maybe a way to override escaping per template. So you'd probably need to specify the return type and 'macro magic' it anyway.

I'll think about it and play with some code.

**Comment by kevinresol** [comment]
I just hit the same thing today.

Isn't it simpler to specify the return type of a function? e.g.
`(: static function renderHtml():Html :)`
`(: static function renderPlain():String :)`

So you can mix different renders in the same template class.

**Comment by back2dos** [comment]
Implemented Kevin's suggestion in version 0.4.2.
